### PR TITLE
Consider off/on in addition to duration when becoming consistent

### PR DESCRIPTION
**Pull request by michaelkirk-signal**
Fixes https://github.com/signalapp/Signal-iOS/issues/3365

PTAL @charlesmchen

### CODE REVIEW

File: SignalServiceKit/src/Messages/OWSDisappearingMessagesJob.m [hunk 1]
```diff
@@ -225,27 +226,25 @@ - (void)becomeConsistentWithConfigurationForMessage:(TSMessage *)message
...
     if (message.expiresInSeconds == 0) {
-        if (disappearingMessagesConfiguration.isEnabled) {
-            changed = YES;
-            DDLogWarn(@"%@ Received remote message which had no expiration set, disabling our expiration to become "
-                      @"consistent.",
-                      self.logTag);
-            disappearingMessagesConfiguration.enabled = NO;
-            [disappearingMessagesConfiguration saveWithTransaction:transaction];
-        }
-    } else if (message.expiresInSeconds != disappearingMessagesConfiguration.durationSeconds) {
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Messages/OWSDisappearingMessagesJob.m]: Previously we were only considering duration and not "enabled" state when becoming consistent with disappearing messages.

File: SignalServiceKit/src/Messages/OWSDisappearingMessagesJob.m [hunk 2]
```diff
@@ -225,27 +226,25 @@ - (void)becomeConsistentWithConfigurationForMessage:(TSMessage *)message
...
+        return;
+    }
+    DDLogInfo(@"%@ becoming consistent message configuration: %@",
+        self.logTag,
+        disappearingMessagesConfiguration.dictionaryValue);
+    [disappearingMessagesConfiguration saveWithTransaction:transaction];
+
+    TSThread *_Nullable thread = [message threadWithTransaction:transaction];
+    if (!thread) {
+        // If there's not yet a message thread, we don't create one.
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Messages/OWSDisappearingMessagesJob.m]: This isn't a change in behavior, I'm just making existing behavior clearer *and* using the transaction we already have vs the sneaky transaction in `message.thread`.

File: SignalServiceKit/src/Messages/OWSDisappearingMessagesJob.m [hunk 3]
```diff
@@ -225,27 +226,25 @@ - (void)becomeConsistentWithConfigurationForMessage:(TSMessage *)message
...
+        disappearingMessagesConfiguration.enabled = NO;
+    } else {
         disappearingMessagesConfiguration.enabled = YES;
         disappearingMessagesConfiguration.durationSeconds = message.expiresInSeconds;
-        [disappearingMessagesConfiguration saveWithTransaction:transaction];
     }
-    
-    if (!changed) {
+
+    if (!disappearingMessagesConfiguration.dictionaryValueDidChange) {
```
Comment (charlesmchen-signal) [inline SignalServiceKit/src/Messages/OWSDisappearingMessagesJob.m]: Will `dictionaryValueDidChange` correctly handle the "create" case of `fetchOrCreateDefaultWithThreadId:...`

File: SignalServiceKit/src/Messages/OWSDisappearingMessagesJob.m [hunk 4]
```diff
@@ -225,27 +226,25 @@ - (void)becomeConsistentWithConfigurationForMessage:(TSMessage *)message
...
+        return;
+    }
+    DDLogInfo(@"%@ becoming consistent message configuration: %@",
+        self.logTag,
+        disappearingMessagesConfiguration.dictionaryValue);
+    [disappearingMessagesConfiguration saveWithTransaction:transaction];
+
+    TSThread *_Nullable thread = [message threadWithTransaction:transaction];
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Messages/OWSDisappearingMessagesJob.m]: If it's a thread created within this transaction, the sneaky transaction in message.thread won't see it, so the initial info message gets lost.

**Pull request merged**